File: golang/yeelight/cache.go
package yeelight

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

const cacheFileName = ".yeelight-cli-cache.json"

type discoveryCacheFile struct {
	CachedAt int64              `json:"cachedAt"`
	Devices  []DiscoveredDevice `json:"devices"`
}

func GetCachePath() string {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		return cacheFileName
	}
	return filepath.Join(home, cacheFileName)
}

func ReadDiscoveryCache(ttlMS int) ([]DiscoveredDevice, bool, error) {
	if ttlMS <= 0 {
		ttlMS = DefaultCacheTTL
	}

	raw, err := os.ReadFile(GetCachePath())
	if err != nil {
		if os.IsNotExist(err) {
			return nil, false, nil
		}
		return nil, false, err
	}

	var parsed discoveryCacheFile
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, false, fmt.Errorf("Discovery cache at %s is not valid JSON.", GetCachePath())
	}

	if time.Now().UnixMilli()-parsed.CachedAt > int64(ttlMS) {
		return nil, false, nil
	}

	return parsed.Devices, true, nil
}

func WriteDiscoveryCache(devices []DiscoveredDevice) error {
	payload := discoveryCacheFile{
		CachedAt: time.Now().UnixMilli(),
		Devices:  devices,
	}
	raw, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(GetCachePath(), raw, 0o644)
}

File: src/yeelight/cache.js
const fs = require("node:fs/promises");
const os = require("node:os");
const path = require("node:path");
const { DEFAULT_CACHE_TTL_MS } = require("./protocol");

const CACHE_PATH = path.join(os.homedir(), ".yeelight-cli-cache.json");

async function readDiscoveryCache(options = {}) {
  const ttlMs = options.ttlMs ?? DEFAULT_CACHE_TTL_MS;
  let raw;

  try {
    raw = await fs.readFile(CACHE_PATH, "utf8");
  } catch (error) {
    if (error && error.code === "ENOENT") {
      return null;
    }

    throw error;
  }

  let parsed;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    throw new Error(`Discovery cache at ${CACHE_PATH} is not valid JSON.`);
  }

  if (!parsed || !Array.isArray(parsed.devices) || typeof parsed.cachedAt !== "number") {
    throw new Error(`Discovery cache at ${CACHE_PATH} has an unexpected shape.`);
  }

  if (Date.now() - parsed.cachedAt > ttlMs) {
    return null;
  }

  return parsed.devices;
}

async function writeDiscoveryCache(devices) {
  const payload = JSON.stringify(
    {
      cachedAt: Date.now(),
      devices
    },
    null,
    2
  );

  await fs.writeFile(CACHE_PATH, payload, "utf8");
}

function getCachePath() {
  return CACHE_PATH;
}

module.exports = {
  getCachePath,
  readDiscoveryCache,
  writeDiscoveryCache
};
